README.md
# csv-clj

Write CSV files 10x faster than clojure/data.csv and work 
easily with structured data. Built on top of the excellent 
[osiegmar/FastCSV](https://github.com/osiegmar/FastCSV) Java library.

v0.0.2 | [CHANGELOG](CHANGELOG.md) | Uses [Break Versioning](https://github.com/ptaoussanis/encore/blob/master/BREAK-VERSIONING.md)

## Usage

The basic read / write interface is nearly a drop-in replacement for
`data.csv` (only the writer construction changes):
```clojure
;; The basic read / write interface is nearly a drop-in replacement for
;; `data.csv` (only the writer construction changes):
(require '[eth.mjd.csv-clj :as csv]
         '[clojure.java.io :as io])

(with-open [writer (csv/writer (io/writer "out-file.csv"))]
  (csv/write-csv writer
                 [["abc" "def"]
                  ["ghi" "jkl"]]))

(with-open [reader (csv/reader (io/reader "out-file.csv"))]
  (doall (csv/read-csv reader)))
;=> (["abc" "def"] ["ghi" "jkl"])
```

## Benchmarks

Benchmarks for writing and then reading 1 million rows:

|                    | csv-clj | clojure.data/csv | Faster by |
|--------------------|---------|------------------|-----------|
| Write unstructured |   236ms |           3004ms |    12.73x |
|   Write structured |   210ms |           2940ms |    14.00x |
|  Read unstructured |   324ms |           1422ms |     4.39x |
|    Read structured |   553ms |           1873ms |     3.39x |


Run with `clj -X:bench`.

## Structured data

The interface for structured data helps with two-way conversion of map or 
record fields, and is also very fast. E.g. for some data where the :timestamp 
of each row is written as an ISO-8601 string, and parsed back in as if via the
`#inst` reader macro.

```clojure
(def data
  [{:timestamp #inst"2021-11-01T19:23:15"
    :double 0.123456789101112
    :description "well, that's a precise number"}
   {:timestamp #inst"2021-11-01T19:23:14"
    :double 0.123456789101111
    :description "this row is only slightly different"}])

; Define the conversion on the way out
(import '(java.time.format DateTimeFormatter) (java.time Instant))
(defn pr-iso-8601 [time-inst]
  (let [inst (Instant/ofEpochMilli (inst-ms time-inst))]
    (.format DateTimeFormatter/ISO_INSTANT inst)))

; Define a codec for the out and in mapping fns for each CSV row
(def codec
  (csv/codec
    [:timestamp pr-iso-8601 clojure.instant/read-instant-date]
    [:double str #(Double/parseDouble %)]
    [:description str str]))

(with-open [writer (csv/writer (io/writer "out-file.csv"))]
  (csv/write-csv-with codec writer data))

(with-open [reader (csv/reader (io/reader "out-file.csv"))]
  (first (csv/read-csv-with reader codec)))
;=> {:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}
```

You can also use the codec directly to convert to and from rows:

```clojure
(csv/>row codec (first data))
;=> ["2021-11-01T19:23:15Z" "0.123456789101112" "well, that's a precise number"]

(csv/<row codec ["2021-11-01T19:23:15Z" "0.123456789101112" "well, that's a precise number"])
;=> {:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}
```

Finally, for peak performance, you can represent rows with records instead of 
maps:
```clojure
(defrecord CSVRow [timestamp double description])

(with-open [reader (csv/reader (io/reader "out-file.csv"))]
  (first
    (csv/read-csv-with
      reader
      (csv/record-codec
        CSVRow
        [:timestamp pr-iso-8601 clojure.instant/read-instant-date]
        [:double str #(Double/parseDouble %)]
        [:description str str]))))
;=> #CSVRow{:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}
```


CHANGELOG.md
# Changelog

## v0.0.2
- Allow reading and writing structured CSVs without headers.

## v0.0.1

- Unit tests.
- Realization that fast record field access is not a good idea when field names
  conflict with record methods (i.e. `(= (.foo x) (:foo x))` but 
  `(not= (.count x) (:count x))`) and subsequent fix.
- Initial functionality and benchmarks.


test/eth/mjd/csv_clj_test.clj
(ns eth.mjd.csv-clj-test
  (:require [clojure.test :refer :all]
            [eth.mjd.csv-clj :as csv]
            [clojure.instant :as instant])
  (:import (java.time Instant)
           (java.time.format DateTimeFormatter)))

(def unstructred-csv
  (str "2021-11-14,plants,3,\"one fewer\"\n"
       "2021-11-13,cats,2,\"active, meowing, \neating houseplants\"\n"
       "2021-11-13,plants,4"))

(def structured-csv
  (str "timestamp,subject,count,description\n"
       "2021-11-14T00:00:00Z,plants,3,one fewer\n"
       "2021-11-13T00:00:00Z,cats,2,\"active, meowing,\neating houseplants\"\n"
       "2021-11-13T00:00:00Z,plants,4,\n"))

(defn pr-iso-8601
  "Print an inst as an ISO-8601 string."
  [time-inst]
  (let [inst (Instant/ofEpochMilli (inst-ms time-inst))]
    (.format DateTimeFormatter/ISO_INSTANT inst)))

;; Nb that `count` is meant to test names conflicting with a record field name.
;; I.e. (.count record) returns something different than (:count record)
(defrecord StructuredRow [timestamp subject count description])
(def codec
  (csv/codec
    [:timestamp pr-iso-8601 instant/read-instant-date]
    [:subject name keyword]
    [:count str #(Integer/parseInt %)]
    [:description identity identity]))
(def rcodec
  (csv/record-codec
    StructuredRow
    [:timestamp pr-iso-8601 instant/read-instant-date]
    [:subject name keyword]
    [:count str #(Integer/parseInt %)]
    [:description identity identity]))

(deftest reading
  (testing "reading unstructured data"
    (let [csv (csv/read-csv unstructred-csv)]
      (is (= (count csv) 3))
      (is (= (second csv) ["2021-11-13" "cats" "2" "active, meowing, \neating houseplants"]))))

  (testing "reading structured data"
    (let [csv (csv/read-csv-with structured-csv codec)]
      (is (= (count csv) 3))
      (is (= (second csv)
             {:timestamp #inst"2021-11-13T00:00:00.000-00:00",
              :subject :cats,
              :count 2,
              :description "active, meowing,\neating houseplants"})))

    (testing "into a record"
      (let [csv (csv/read-csv-with structured-csv rcodec)]
        (is (= (count csv) 3))
        (is (= (second csv)
               (->StructuredRow
               #inst"2021-11-13T00:00:00.000-00:00",
                :cats
                2
                "active, meowing,\neating houseplants")))))))

(deftest writing
  (testing "writing unstructured data"
    (let [row-vectors (csv/read-csv unstructred-csv)]
      (is (= (-> row-vectors csv/write-csv csv/read-csv) row-vectors))))
  (testing "writing structured data"
    (let [row-maps (csv/read-csv-with structured-csv codec)
          written (csv/write-csv-with codec row-maps)]
      (is (= (csv/read-csv-with written codec) row-maps)))
    (testing "from records"
      (let [row-records (csv/read-csv-with structured-csv rcodec)
            written (csv/write-csv-with rcodec row-records)]
        (is (= (csv/read-csv-with written rcodec) row-records))))))


bench/bench.clj
(ns bench
  "Read and write some data, both with this library and with clojure.data/csv,
  and print the results in a table.

  Invoke with `clj -X:bench`, or `clj -X:bench :n 100000` to restrict the
  number of rows used."
  (:require [eth.mjd.csv-clj :as csv]
            [clojure.data.csv :as data.csv]

            [clojure.java.io :as io]
            [clojure.pprint :as pprint]
            [clojure.string :as string])
  (:import (java.util Random)
           (java.io ByteArrayOutputStream ByteArrayInputStream)))

(set! *warn-on-reflection* true)


;;; Data used for benchmarking
(def max-rows 10000000)
(def alphabet (vec (concat [\" \,] (map char (range 58 97)))))

(def unstructured-data
  (let [r (Random. 1635792327272)]
    (for [_ (range max-rows)]
      (let [row-len (inc (.nextInt r 8))]
        (vec
          (for [_ (range row-len)]
            (let [cell-len (.nextInt r 8)]
              (apply str
                     (for [_ (range cell-len)]
                       (nth alphabet (.nextInt r (count alphabet))))))))))))

(defrecord BenchmarkRow [foo bar baz qux])
(def structured-data
  (let [r (Random. 1635792327272)]
    (for [_ (range max-rows)]
      (let [cell #(let [cell-len (.nextInt r 8)]
                    (apply str
                           (for [_ (range cell-len)]
                             (nth alphabet (.nextInt r (count alphabet))))))]
        (->BenchmarkRow (cell) (cell) (cell) (cell))))))


;;; Simple benchmark

(def unstructured-bytes (atom nil))
(def structured-bytes (atom nil))
(def ^:const cap (long (Math/pow 2 28)))

(defn clojure-data-csv-unstructured-write [data]
  (with-open [w (io/writer (ByteArrayOutputStream. cap))]
    (data.csv/write-csv w data)))

(defn clojure-data-csv-unstructured-read []
  (with-open [r (io/reader (ByteArrayInputStream. @unstructured-bytes))]
    (doall (data.csv/read-csv r))))

(defn unstructured-write [data]
  (let [baos (ByteArrayOutputStream. cap)]
    (with-open [w (csv/writer (io/writer baos))]
      (csv/write-csv w data)
      (reset! unstructured-bytes (.toByteArray baos)))))

(defn unstructured-read []
  (with-open [r (csv/reader (io/reader (ByteArrayInputStream. @unstructured-bytes)))]
    (doall (csv/read-csv r))))


;;; Benchmark with structured data

(defn clojure-data-csv-structured-write [data]
  (with-open [w (io/writer (ByteArrayOutputStream. cap))]
    (let [x (first data)]
      (data.csv/write-csv
        w
        (cons
          (map name (keys x)) ; headers
          (map vals data)))))) ; rows

(defn clojure-data-csv-structured-read []
  (with-open [r (io/reader (ByteArrayInputStream. @structured-bytes))]
    (let [data (data.csv/read-csv r)
          headers (map keyword (first data))]
      (->> (rest data)
           (map #(zipmap headers %))
           doall))))

(def codec
  (csv/record-codec
    BenchmarkRow
    [:foo identity identity]
    [:bar identity identity]
    [:baz identity identity]
    [:qux identity identity]))

(defn structured-write [data]
  (let [baos (ByteArrayOutputStream. cap)]
    (with-open [w (csv/writer (io/writer baos))]
      (csv/write-csv-with w codec data))
    (reset! structured-bytes (.toByteArray baos))))

(defn structured-read []
  (with-open [r (csv/reader (io/reader (ByteArrayInputStream. @structured-bytes)))]
    (doall (csv/read-csv-with r codec))))


;;; Code to run the benchmarks

(defn median [nums]
  (loop [nums (vec (sort nums))]
    (let [cnt (count nums)]
      (cond
        (= cnt 1) (first nums)
        (= cnt 2) (/ (apply + nums) 2.0)
        :else (recur (subvec nums 1 (dec cnt)))))))

(defn bench* [dbg f trials]
  (System/gc)

  (print "Running warm up iterations: ")
  (flush)
  (dotimes [i 4]
    (f)
    (print i "")
    (flush))
  (println "")

  (let [times (mapv
                (fn [trial]
                  (print (format "#%-2s Timing %s... " trial dbg))
                  (flush)
                  (let [start (System/currentTimeMillis)]
                    (f)
                    (let [elapsed (- (System/currentTimeMillis) start)]
                      (print elapsed)
                      (println "msecs")
                      elapsed)))
                (range trials))
        m (median times)]
    (println (format "Median time for %s is %smsecs\n" dbg m))
    m))

(defmacro bench [f trials & [?data]]
  (if ?data
    `(let [data# ~?data]
       (bench* ~(str f) (fn [] (~f data#)) ~trials))
    `(bench* ~(str f) ~f ~trials)))

(def ^:const trials 5)

(defn write-benchmarks [n]
  (io/make-parents "bench/test.csv")
  (let [_ (println "Generating benchmark data...")
        udata (into [] (take n) unstructured-data)
        sdata (into [] (take n) structured-data)
        tbl [{""                 "Write unstructured"
              "csv-clj"          (str (bench unstructured-write trials udata) "ms")
              "clojure.data/csv" (str (bench clojure-data-csv-unstructured-write trials udata) "ms")}
             {""                 "Write structured"
              "csv-clj"          (str (bench structured-write trials sdata) "ms")
              "clojure.data/csv" (str (bench clojure-data-csv-structured-write trials sdata) "ms")}]]
    tbl))

(defn read-benchmarks []
  (let [tbl [{""                 "Read unstructured"
              "csv-clj"          (str (bench unstructured-read trials) "ms")
              "clojure.data/csv" (str (bench clojure-data-csv-unstructured-read trials) "ms")}
             {""                 "Read structured"
              "csv-clj"          (str (bench structured-read trials) "ms")
              "clojure.data/csv" (str (bench clojure-data-csv-structured-read trials) "ms")}]]
    tbl))

(defn benchmark [& {:keys [n] :or {n 1000000}}]
  (assert (<= n max-rows))
  (let [nstr (apply
               format
               (cond
                 (>= n 1000000) ["%smm" (/ n 1000000.0)]
                 (>= n 1000) ["%sk" (/ n 1000.0)]
                 :else ["%s" n]))
        _ (println "Benchmarking with" nstr "rows of data...")
        wbs (write-benchmarks n)
        rbs (read-benchmarks)
        comparison (fn [x]
                     (let [this (get x "csv-clj")
                           that (get x "clojure.data/csv")
                           n (fn [s] (-> (string/split s #"m") first bigdec double))]
                       (assoc x "Faster by" (format "%.2fx" (/ (n that) (max (n this) 1))))))]
    (println "Benchmarks for writing and then reading" nstr "rows:")
    (pprint/print-table (map comparison (concat wbs rbs)))))

(comment
  ;; Run like so
  (benchmark)
  (benchmark :n 200000)
  (benchmark :n 1000000))


src/eth/mjd/csv_clj.clj
(ns eth.mjd.csv-clj
  "Fast CSV reading and writing for both structured and unstructured data."
  (:require [clojure.java.io :as io])
  (:refer-clojure :exclude [read write])
  (:import (de.siegmar.fastcsv.writer CsvWriter LineDelimiter)
           (java.io Closeable StringReader StringWriter)
           (de.siegmar.fastcsv.reader CsvReader CloseableIterator CsvRow)
           (java.lang.reflect Method)))

(set! *warn-on-reflection* true)

;;; Base CSV reading / writing functionality

(defprotocol ICSVWriter (write [this cells] "Write a row of [string]."))
(defprotocol ICSVReader (read [this] "Read a row of [string]."))

(deftype CSVWriter [^CsvWriter delegate-writer]
  ICSVWriter (write [this cells] (.writeRow delegate-writer ^Iterable cells))
  Closeable (close [this] (.close delegate-writer)))

(deftype CSVReader [^CloseableIterator rows]
  ICSVReader (read [this] (when (.hasNext rows) (.getFields ^CsvRow (.next rows))))
  Closeable (close [this] (.close rows)))

(defn ^CSVWriter writer
  "Return a CSV writer for the given file, path, or backing writer.

  Should be used inside `with-open` to ensure that the writer is properly
  closed.

  Valid options are
    :separator (Default \\,)
    :quote (Default \\\")
    :newline (:lf (default) or :cr+lf)"
  ([x] (writer x {}))
  ([x {:keys [separator quote newline]}]
   (CSVWriter.
     (.build
       (cond-> (CsvWriter/builder)
               separator (.fieldSeparator separator)
               quote (.quoteCharacter quote)
               newline (.lineDelimiter (case newline
                                         :lf LineDelimiter/LF
                                         :cr+lf LineDelimiter/CRLF)))
       (io/writer x)))))

(defn ^CSVReader reader
  "Return a CSV reader for the given file, path, or backing reader.

  Should be used inside `with-open` to ensure that the reader is properly
  closed.

  Valid options are
    :separator (Default \\,)
    :quote (Default \\\")"
  ([x] (reader x {}))
  ([x {:keys [separator quote]}]
   (CSVReader.
     (-> (cond-> (CsvReader/builder)
                 separator (.fieldSeparator separator)
                 quote (.quoteCharacter quote))
         (.build (io/reader x))
         (.iterator)))))

(defn write-csv
  "Write rows (shaped [[string]]) with the given writer (eager)."
  ([rows]
   (let [sw (StringWriter.)]
     (with-open [w (writer sw)]
       (write-csv w rows))
     (.toString sw)))
  ([csv-writer rows]
   (run! (fn [row] (write csv-writer row)) rows)))

(defn read-csv
  "Read rows (shaped [[string]]) with the given reader (lazy)."
  [csv-reader]
  (if (string? csv-reader)
    (read-csv (reader (StringReader. csv-reader)))
    (take-while some? (repeatedly (fn [] (read csv-reader))))))

(comment
  ; Writing
  (with-open [w (writer "out.csv")]
    (write-csv w [["some" "headers" "to" "write"]
                  ["foo" "bar" "baz" "qux"]]))

  ; Reading
  (with-open [r (reader "out.csv")]
    (doall (read-csv r))))


;;; Codecs for working with structured data

(defprotocol TableCodec
  (headers [this] "The headers for a the table (ordered).")
  (>row [this x] "Transform to a [string] row.")
  (<row [this x] "Transform from a [string] row."))

(defmacro codec
  "Compile a codec for reading and writing tabular CSV data.

  Takes a series of `[field-name, out-fn, in-fn]` specifications and builds
  an encoding function which transform a data point to a vector of
  [(out-fn (get data field)) ...], with each vector item corresponding to one
  of the fields, and does the inverse for the decoding function.

  E.g. for a table with elements represented by maps shaped

      {:timestamp 1635812096876
       :price 100.99M
       :market :eth-usd}

  You might compile a codec:

      (codec
        [:timestamp str #(Long/parseLong %)]
        [:price #(.toPlainString ^BigDecimal %) bigdec]
        [:market name keyword])"
  [& fields+outf+inf]
  (let [sym (gensym)]
    `(reify TableCodec
       (headers [_#]
         [~@(mapv
              (comp #(if (keyword? %) (name %) (str %)) first)
              fields+outf+inf)])
       (>row [_# ~sym]
         [~@(map
              (fn [[field outf]]
                (if (= outf `identity)
                  (list `get sym field)
                  (list outf (list `get sym field))))
              fields+outf+inf)])
       (<row [_# ~sym]
         (array-map
           ~@(mapcat
               (fn [[idx [field _ inf]]]
                 [field
                  (if (= inf `identity)
                    (list `nth sym idx)
                    (list inf (list `nth sym idx)))])
               (map-indexed vector fields+outf+inf)))))))


;; E.g. a record field named "size" can't be accessed via (.size record),
;; because "size" already refers to a method.
(defrecord Dummy [])
(defonce forbidden-names (into #{} (map #(.getName ^Method %) (.getMethods Dummy))))


(defmacro record-codec
  "Like `codec`, but takes a record type as the first argument.

  When reading a row, the parsed contents at passed to the record constructor
  *in the order that the fields are passed into this function* to construct the
  codec.

  E.g. you might

      (defrecord MyRow [timestamp price market])
      (record-codec
        MyRow
        [:timestamp str #(Long/parseLong %)]
        [:price #(.toPlainString ^BigDecimal %) bigdec]
        [:market name keyword])"
  [record-type & fields+outf+inf]
  (let [sym (gensym)]
    `(reify TableCodec
       (headers [_#]
         [~@(mapv
              (comp #(if (keyword? %) (name %) (str %)) first)
              fields+outf+inf)])
       (>row [_# ~sym]
         [~@(map
              (fn [[field outf]]
                (assert (keyword? field) "all record fields are keywords")
                ; Some fields can't be accessed via .field
                (let [get-field (if (contains? forbidden-names (name field))
                                  (list field sym)
                                  (list `..
                                        (with-meta sym {:tag record-type})
                                        (symbol (name field))))]
                  (if (= outf `identity)
                    get-field
                    (list outf get-field))))
              fields+outf+inf)])
       (<row [_# ~sym]
         (new ~record-type
           ~@(map
               (fn [[idx [_ _ inf]]]
                 (if (= inf `identity)
                   (list `nth sym idx)
                   (list inf (list `nth sym idx)))) ;; TODO: Is nth slow?
               (map-indexed vector fields+outf+inf)))))))


(comment
  ;; E.g.
  (defrecord Foo [x size])
  (macroexpand-1 `(record-codec Foo [:x identity identity] [:size str bigdec])))

(defn write-csv-with
  "Write a CSV from `data`, a series of maps / records, and a `codec`.

  Pass {:omit-header? true} to write the rows only, without headers."
  ([codec data]
   (let [sw (StringWriter.)]
     (with-open [w (writer sw)]
       (write-csv-with w codec data))
     (.toString sw)))
  ([writer codec data & {:keys [omit-header?]}]
   (when-not omit-header?
     (write writer (headers codec)))
   (run! (fn [row] (write writer (>row codec row))) data)))

(defn read-csv-with
  "Read a CSV into a series of maps or records according to `codec`.

  Pass {:omit-header? true} if the CSV contains no headers, in which case the
  first row interpreted positionally without checking that the headers match
  the codec."
  [csv-reader codec & {:keys [omit-header?]}]
  (if (string? csv-reader)
    (read-csv-with (reader (StringReader. csv-reader)) codec)

    (do
      ; discard headers -- codec already knows what they should be
      (when-not omit-header?
        (let [hd (read csv-reader)]
          (assert
            (= hd (headers codec))
            (format "headers (%s) match codec (%s)" (vec hd) (headers codec)))))

      (sequence
        (comp
          (take-while some?)
          (map #(<row codec %)))
        (repeatedly (fn [] (read csv-reader)))))))

;;; Examples

(comment
  ;; The basic read / write interface is nearly a drop-in replacement for
  ;; `data.csv` (only the writer construction changes):
  (require '[eth.mjd.csv-clj :as csv]
           '[clojure.java.io :as io])

  (with-open [writer (csv/writer (io/writer "out-file.csv"))]
    (csv/write-csv writer
                   [["abc" "def"]
                    ["ghi" "jkl"]]))

  (with-open [reader (csv/reader (io/reader "out-file.csv"))]
    (doall (csv/read-csv reader))))

(comment
  ;; The interface for typed data helps with two-way conversion of map or record
  ;; fields, and is also very fast. E.g. for some data where the :timestamp of
  ;; each row is written as an ISO-8601 string, and parsed back in as if via the
  ;; #inst reader macro.
  (def data
    [{:timestamp   #inst"2021-11-01T19:23:15"
      :double      0.123456789101112
      :description "well, that's a precise number"}
     {:timestamp   #inst"2021-11-01T19:23:14"
      :double      0.123456789101111
      :description "this row is only slightly different"}])

  ; Define the conversion on the way out
  (import '(java.time.format DateTimeFormatter) (java.time Instant))
  (defn pr-iso-8601 [time-inst]
    (let [inst (Instant/ofEpochMilli (inst-ms time-inst))]
      (.format DateTimeFormatter/ISO_INSTANT inst)))

  ; Define a codec for the out and in mapping fns for each CSV row
  (def codec
    (csv/codec
      [:timestamp pr-iso-8601 clojure.instant/read-instant-date]
      [:double str #(Double/parseDouble %)]
      [:description str str]))

  (with-open [writer (csv/writer (io/writer "out-file.csv"))]
    (csv/write-csv-with writer codec data))

  (with-open [reader (csv/reader (io/reader "out-file.csv"))]
    (first (csv/read-csv-with reader codec)))
  ;=> {:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}
  )

(comment
  ;; You can also use the codec directly to convert to and from rows
  (csv/>row codec (first data))
  ;=> ["2021-11-01T19:23:15Z" "0.123456789101112" "well, that's a precise number"]

  (csv/<row codec ["2021-11-01T19:23:15Z" "0.123456789101112" "well, that's a precise number"])
  ;=> {:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}

  ;; Finally, for performance, you can use records instead of maps:
  (defrecord CSVRow [timestamp double description])

  (with-open [reader (csv/reader (io/reader "out-file.csv"))]
    (first
      (csv/read-csv-with
        reader
        (csv/record-codec
          CSVRow
          [:timestamp pr-iso-8601 clojure.instant/read-instant-date]
          [:double str #(Double/parseDouble %)]
          [:description str str]))))
  ;=> #CSVRow{:timestamp #inst"2021-11-01T19:23:15.000-00:00", :double 0.123456789101112, :description "well, that's a precise number"}
  )
